app/scripts/metamask-controller.js

```diff
@@ -2204,12 +2207,65 @@ export default class MetamaskController extends EventEmitter {
   }
 
   async getTokenStandardAndDetails(address, userAddress, tokenId) {
-    const details =
-      await this.assetsContractController.getTokenStandardAndDetails(
+    const { tokenList } = this.tokenListController.state;
+    const { tokens } = this.tokensController.state;
+
+    const staticTokenListDetails =
+      STATIC_MAINNET_TOKEN_LIST[address.toLowerCase()] || {};
+    const tokenListDetails = tokenList[address.toLowerCase()] || {};
+    const userDefinedTokenDetails =
+      tokens.find(({ address: _address }) =>
+        isEqualCaseInsensitive(_address, address),
+      ) || {};
+
+    const tokenDetails = {
+      ...staticTokenListDetails,
+      ...tokenListDetails,
+      ...userDefinedTokenDetails,
+    };
+    const tokenDetailsStandardIsisERC20 =
+      isEqualCaseInsensitive(tokenDetails.standard, TokenStandard.ERC20) ||
+      tokenDetails.erc20 === true;
+    const noEvidenceThatTokenIsAnNFT =
+      !isEqualCaseInsensitive(tokenDetails.standard, TokenStandard.ERC1155) &&
+      !isEqualCaseInsensitive(tokenDetails.standard, TokenStandard.ERC721) &&
+      !tokenDetails.erc721;
+    const otherThanStandardDetailsAreERC20Like =
+      tokenDetails.decimals !== undefined && tokenDetails.symbol;
+
+    const tokenCanBeTreatedAsAnERC20 =
+      tokenDetailsStandardIsisERC20 ||
+      (noEvidenceThatTokenIsAnNFT && otherThanStandardDetailsAreERC20Like);
+
+    let details;
+    if (tokenCanBeTreatedAsAnERC20) {
+      try {
+        const balance = await fetchTokenBalance(
+          address,
+          userAddress,
+          this.provider,
+        );
+
+        details = {
+          address,
+          balance,
+          standard: TokenStandard.ERC20,
+          decimals: tokenDetails.decimals,
+          symbol: tokenDetails.symbol,
+        };
+      } catch (e) {
+        log.warning('Failed to get token balance');
+      }
+    }
+
+    if (details === undefined) {
```

Comment:
same functionality and easier to read/understand imo: 

```suggestion
    if (tokenCanBeTreatedAsAnERC20) {
      try {
        const balance = await fetchTokenBalance(
          address,
          userAddress,
          this.provider,
        );

        details = {
          address,
          balance,
          standard: TokenStandard.ERC20,
          decimals: tokenDetails.decimals,
          symbol: tokenDetails.symbol,
        };
      } catch (e) {
        log.warning('Failed to get token balance');
      }
    } else {
```

Reply:
So this is not exactly the same, but your comment shows that my code is hard to understand, and at the very least requires some serious documentation.

The current code will go into the `if (details === undefined) {` block if it first entered `if (tokenCanBeTreatedAsAnERC20) {` but then `const balance = await fetchTokenBalance(`. So it tries to fetch the balance, but if that fails, tries to get the data via the assets controller call.

The suggested change here will not try the assets controller call if the `fetchTokenBalance` call fails.

Reply:
oh true. Wasn't thinking straight ! Yeah maybe just a short comment here would help. I don't think it's that bad though.

Reply:
clarifying comments added in a92ffd62d5

Reply:
I think I preferred the old name `tokenCanBeTreatedAsAnERC20` over `weCanAttemptToFetchERC20TokenBalance` because, ERC721 and ERC1155 both have a `balanceOf` method

Reply:
okay, good point

Reply:
👍